tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -75,45 +75,149 @@ StorageManager::~StorageManager() {
   delete tile_cache_;
   delete vfs_;
 
-  // Delete all open arrays, force-releasing any filelocks
-  open_array_mtx_.lock();
+  // Release all filelocks
   for (auto& open_array_it : open_arrays_) {
     open_array_it.second->file_unlock(vfs_);
     delete open_array_it.second;
   }
-  open_array_mtx_.unlock();
+
+  for (auto& fl_it : xfilelocks_) {
+    auto filelock = fl_it.second;
+    auto lock_uri = URI(fl_it.first).join_path(constants::filelock_name);
+    if (filelock != INVALID_FILELOCK)
+      vfs_->filelock_unlock(lock_uri, filelock);
+  }
 }
 
 /* ****************************** */
 /*               API              */
 /* ****************************** */
 
+Status StorageManager::array_close(const URI& array_uri) {
+  // Lock mutex
+  open_array_mtx_.lock();
+
+  // Find the open array entry
+  auto it = open_arrays_.find(array_uri.to_string());
+
+  // Do nothing if array is closed
+  if (it == open_arrays_.end()) {
+    open_array_mtx_.unlock();
+    return Status::Ok();
+  }
+
+  // For easy reference
+  OpenArray* open_array = it->second;
+
+  // Lock the mutex of the array and decrement counter
+  open_array->mtx_lock();
+  open_array->cnt_decr();
+
+  // Close the array if the counter reaches 0
+  if (open_array->cnt() == 0) {
+    // Release file lock
+    auto st = open_array->file_unlock(vfs_);
+    if (!st.ok()) {
+      open_array->mtx_unlock();
+      open_array_mtx_.unlock();
+      return st;
+    }
+
+    // Remove open array entry
+    open_array->mtx_unlock();
+    delete open_array;
+    open_arrays_.erase(it);
+  } else {  // Just unlock the array mutex
+    open_array->mtx_unlock();
+  }
+
+  // Unlock mutex and notify condition on exclusive locks
+  open_array_mtx_.unlock();
+  xlock_cv_.notify_all();
+
+  return Status::Ok();
+}
+
+Status StorageManager::array_open(
+    const URI& array_uri, OpenArray** open_array) {
+  // Check if array exists
+  ObjectType obj_type;
+  RETURN_NOT_OK(this->object_type(array_uri, &obj_type));
+  if (obj_type != ObjectType::ARRAY && obj_type != ObjectType::KEY_VALUE) {
+    return LOG_STATUS(
+        Status::StorageManagerError("Cannot open array; Array does not exist"));
+  }
+
+  // Lock mutex
+  open_array_mtx_.lock();
+
+  // Find the open array entry
+  auto it = open_arrays_.find(array_uri.to_string());
+  if (it != open_arrays_.end()) {
+    *open_array = it->second;
+  } else {  // Create a new entry
+    *open_array = new OpenArray(array_uri);
+    open_arrays_[array_uri.to_string()] = *open_array;
+  }
+
+  // Lock the array and increment counter
+  (*open_array)->mtx_lock();
+  (*open_array)->cnt_incr();
+
+  // Unlock mutex
+  open_array_mtx_.unlock();
+
+  // Acquire a shared filelock
+  RETURN_NOT_OK((*open_array)->file_lock(vfs_));
```

Comment:
Do we need `(*open_array)->mtx_unlock()` on the error path here?

Reply:
Yes, will fix.